src/Sylius/Bundle/CoreBundle/Routing/Generator/ProductShopPageUrlGenerator.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Routing\Generator;
+
+use Doctrine\Common\Collections\Collection;
+use Sylius\Component\Core\Model\ChannelInterface;
+use Sylius\Component\Core\Model\ProductInterface;
+use Sylius\Component\Core\Model\ProductTranslationInterface;
+use Sylius\Component\Locale\Context\LocaleContextInterface;
+use Sylius\Component\Locale\Model\LocaleInterface;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+
+final class ProductShopPageUrlGenerator implements ProductShopPageUrlGeneratorInterface
+{
+    public function __construct(
+        private LocaleContextInterface $localeContext,
+        private UrlGeneratorInterface $urlGenerator,
+        private bool $unsecuredUrls,
+    ) {
+    }
+
+    public function generate(ProductInterface $product, ChannelInterface $channel): ?string
+    {
+        /** @var Collection<array-key, ProductTranslationInterface> $productTranslations */
+        $productTranslations = $product->getTranslations();
+
+        $administratorLocaleCode = $this->localeContext->getLocaleCode();
```

Comment:
```suggestion
        $contextLocaleCode = $this->localeContext->getLocaleCode();
```
`administrator` might be misleading since it depends on the priorities